src/notebooksyncagent.cpp

```diff
@@ -782,60 +792,100 @@ bool NotebookSyncAgent::calculateDelta(
+                    // we later check to see if it was modified server-side by checking the etag value.
+                    LOG_DEBUG("have possibly modified or possibly unchanged previously synced local incidence:" << remoteUri);
+                    seenRemoteUris.insert(remoteUri);
+                    previouslySyncedEtags.insert(remoteUri, incidenceETag(incidence));
+                }
             }
         }
     }
 
-    // now determine local deletions.  Note that we combine deletions reported since
-    // the from date and since the last sync date, due to mKCal API semantics.
-    KCalCore::Incidence::List deleted, deletedSyncDate;
-    uniteIncidenceLists(deletedSyncDate, &deleted);
-    if (!mStorage->deletedIncidences(&deleted, fromDate, mNotebook->uid()) ||
-        !mStorage->deletedIncidences(&deletedSyncDate, syncDateTime, mNotebook->uid())) {
+    // now determine local deletions.  Note that we observe deletions reported since
+    // the earlier of the "from date" and "since the last sync date", due to mKCal API semantics.
+    KCalCore::Incidence::List deleted;
+    if (!mStorage->deletedIncidences(&deleted, earliestAnchorTimestamp, mNotebook->uid())) {
         LOG_CRITICAL("mKCal::ExtendedStorage::deletedIncidences() failed");
         return false;
     }
-    Q_FOREACH (KCalCore::Incidence::Ptr incidence, deleted) {
-        bool uriWasEmpty = false;
-        QString remoteUri = incidenceHrefUri(incidence, mRemoteCalendarPath, &uriWasEmpty);
-        if (remoteUriEtags.contains(remoteUri)) {
-            if (uriWasEmpty) {
-                // we originally upsynced this pure-local addition, but then connectivity was
-                // lost before we updated the uid of it locally to include the remote uri.
-                // subsequently, the user deleted the incidence.
-                // Hence, it exists remotely, and has been deleted locally.
-                LOG_DEBUG("have local deletion for partially synced incidence:" << incidence->uid() << incidence->recurrenceId().toString());
-            } else {
-                // the incidence was previously synced successfully.  it has now been deleted locally.
-                LOG_DEBUG("have local deletion for previously synced incidence:" << incidence->uid() << incidence->recurrenceId().toString());
-            }
-            LocalDeletion localDeletion(incidence, remoteUriEtags.value(remoteUri), remoteUri);
-            localDeletions->append(localDeletion);
-            seenRemoteUris.insert(remoteUri);
-        } else {
-            // it was either already deleted remotely, or was never upsynced from the local prior to deletion.
-            LOG_DEBUG("ignoring local deletion of non-existent remote incidence:" << incidence->uid() << incidence->recurrenceId().toString() << "at" << remoteUri);
-        }
+    QSet<QString> deletedSeriesUids;
+    static const int BASE_SERIES = 1, PERSISTENT_EXCEPTION_OCCURRENCES = 0;
```

Comment:
I've refactored this, as I found another case which could cause problems.  Now I only loop over them once here, but later iterate over the set and erase ones which need erasing, AFTER the set of remote modifications is also determined.  On the plus side, the code is nicer too.